bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+				size_t size, bool first)
+{
+	uint16_t *mtid;
+
+	mtid = (uint16_t *)&mtid_tlv->data;
+	for (int i = 0; i < (mtid_tlv->length / 2); i++) {
+		if (i == 0)
+			snprintf(buf, size, "%sMT:%hu", first ? "" : " ",
+				 mtid[i]);
+		else
+			snprintf(buf, size, ",%hu", mtid[i]);
+		size -= strlen(buf);
+		buf += strlen(buf);
+	}
+	return size;
+}
+
+static size_t bgp_linkstate_nlri_link_descriptor_display(
+	struct bgp_ls_nlri_link_descr_tlv *link_descr, char *buf, size_t size,
+	bool multiline)
+{
+	bool first = true;
+
+	if (link_descr->flags == 0)
+		return size;
+
+	snprintf(buf, size, "%sLink%s", multiline ? "\n" : " ",
+		 multiline ? "   " : "{");
+	size -= strlen(buf);
+	buf += strlen(buf);
+
+	if (CHECK_FLAG(link_descr->flags,
+		       BGP_NLRI_TLV_LINK_DESCR_LOCAL_REMOTE_ID)) {
+		snprintf(buf, size, "%sLocal/remote:%hu", first ? "" : " ",
+			 link_descr->local_remote_id);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_INTERFACE4)) {
+		snprintfrr(buf, size, "%sIPv4:%pI4", first ? "" : " ",
+			   &link_descr->interface4);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_NEIGHBOR4)) {
+		snprintfrr(buf, size, "%sNeigh-IPv4:%pI4", first ? "" : " ",
+			   &link_descr->neighbor4);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_INTERFACE6)) {
+		snprintfrr(buf, size, "%sIPv6:%pI6", first ? "" : " ",
+			   &link_descr->interface6);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_NEIGHBOR6)) {
+		snprintfrr(buf, size, "%sNeigh-IPv6:%pI6", first ? "" : " ",
+			   &link_descr->neighbor6);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_MT_ID)) {
+		size = bgp_linkstate_nlri_mtid_display(&link_descr->mtid, buf,
+						       size, first);
+		buf += strlen(buf);
+	}
+
+	if (!multiline) {
+		snprintf(buf, size, "}");
+		size -= strlen(buf);
+	}
+
+	return size;
+}
+
+static uint16_t
+bgp_linkstate_nlri_link_prefixlen(struct bgp_linkstate_type_link *p_link)
+{
+	return sizeof(struct bgp_linkstate_type_link) +
+	       p_link->link_descr.mtid.length;
+}
+
+static int bgp_linkstate_nlri_link_decode(struct bgp_linkstate_type_link **pp,
+					  uint8_t *pnt, uint16_t len)
+{
+	int ret = BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+	struct bgp_linkstate_type_link *p_link = *pp;
+	uint8_t *lim = pnt + len;
+	uint16_t type, length, *dst, *src;
+
+	p_link->proto = pnt_decode8(&pnt);
+	p_link->identifier = pnt_decode64(&pnt);
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+
+		if (!bgp_ls_tlv_check_size(type, length))
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+
```

Comment:
dont return error if unknown tlv.
rfc7752: "[3.1](https://www.rfc-editor.org/rfc/rfc7752#section-3.1). Unrecognized types MUST be preserved and propagated."